[SPARK-5100][SQL] add thriftserver-ui support

tianyi opened the pull request:
In the latest Spark release, there is a Spark Streaming tab on the driver web UI, which shows information about running streaming application. It should be helpful for providing a monitor page in Thrift server because both streaming and Thrift server are long-term applications, and the details of the application do not show on stage page or job page.

Design doc is here: https://issues.apache.org/jira/secure/attachment/12690744/Spark%20Thrift-server%20monitor%20page.pdf

Prototype snapshot: https://issues.apache.org/jira/secure/attachment/12690297/prototype-screenshot.png

tianyi committed:
add thriftserver-ui support

liancheng commented:
ok to test

AmplabJenkins commented:
Can one of the admins verify this patch?

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala:
```diff
@@ -93,6 +94,13 @@ private[hive] class HiveThriftServer2(hiveContext: HiveContext)
   with ReflectedCompositeService {
 
+  private[hive] val uiTab: Option[ThriftServerTab] =
+    if (hiveContext.hiveconf.getBoolean("spark.ui.enabled", true)) {
```
It's a bit weird to read Spark configuration via `HiveConf` although it's valid here. `hiveContext.sparkContext.getConf` might be better.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLEnv.scala:
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.sql.hive.thriftserver
 
+import org.apache.spark.sql.hive.thriftserver.ui.ThriftServerUIEventListener
```
Please reorder imports according to Spark coding conventions.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+
+import scala.collection.mutable.HashMap
+
+trait ThriftServerEventListener {
```
Please limit all classes/traits/objects in this file to `private[sql]`.

liancheng replied:
Also please document all event handler methods, especially at what time each handler is called.

liancheng replied:
Actuall limit to `private[thriftserver]`.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  override def onDisconnected(session: HiveSession): Unit = {
+    if(!sessionList.contains(session.getSessionHandle)) {
```
Space after `if`.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+      onConnected(session)
+    }
+    sessionList(session.getSessionHandle).finishTimestamp = System.currentTimeMillis()
+
```
Remove this empty line.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+  def onConnected(session: HiveSession) { }
+  def onDisconnected(session: HiveSession) { }
+
+  def onStart(id: String, session: HiveSession, statement: String) { }
```
`onQueryStart`?

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+  def onStart(id: String, session: HiveSession, statement: String) { }
+  def onParse(id: String, executePlan: String, groupId: String) { }
+  def onError(id: String, errorMessage: String, errorTrace: String) { }
+  def onFinish(id: String) { }
```
`onQueryFinish`?

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+
+  private def compareExecutionDesc(
+      l:(String, ExecutionInfo),
+      r:(String, ExecutionInfo)): Boolean = {
```
Space after `:`

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+
+  private def compareSessionDesc(
+      l:(SessionHandle, SessionInfo),
+      r:(SessionHandle, SessionInfo)): Boolean = {
```
Space after `:`

liancheng commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suite.scala:
```diff
@@ -121,7 +152,16 @@ class HiveThriftServer2Suite extends FunSuite with Logging {
+    startThriftServerWithUIPort(thriftPort, randomListeningPort, serverStartTimeout, httpMode)(f)
+  }
+
+  def startThriftServerWithUIPort(
```
I think it's OK to reuse the name `startThriftServer ` here.

tianyi replied:
As we discussed, reuse  startThriftServer will cause `multiple overloaded alternatives of method startThriftServer define default arguments`. I will modify the `startThriftServer` function directly.

liancheng commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suite.scala:
```diff
@@ -70,6 +73,34 @@ class HiveThriftServer2Suite extends FunSuite with Logging {
     port
   }
 
+  def withThriftUIAndJDBC(
```
This method duplicates a lot code in `withJdbcStatement`.  It should be easy to eliminate them.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+  }
+}
+
+class ThriftServerUIEventListener(val conf: SparkConf)
```
Actually we've already got a listener in `object HiveThriftServer2`, is it possible to merge them?

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+  override def onError(id: String, errorMessage: String, errorTrace: String): Unit = {
+    executeList(id).finishTimestamp = System.currentTimeMillis()
+    executeList(id).detail = errorMessage
+    //+ "<br></br>" + errorTrace
```
Remove this line

liancheng commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suite.scala:
```diff
@@ -70,6 +73,34 @@ class HiveThriftServer2Suite extends FunSuite with Logging {
+    }
+  }
+
   def withJdbcStatement(
```
We can change `f: Statement => Unit` to `f: (Int, Int, Statement) => Unit`, and pass in server port and UI port. Then `withThriftUIAndJDBC` can be removed.

liancheng commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suite.scala:
```diff
@@ -121,7 +152,16 @@ class HiveThriftServer2Suite extends FunSuite with Logging {
+      thriftPort: Int,
+      serverStartTimeout: FiniteDuration = 1.minute,
+      httpMode: Boolean = false)(
+      f: => Unit): Unit = {
```
Also, change `f: => Unit` to `f: (Int, Int) => Unit` and pass in server port and UI port here.

liancheng commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suite.scala:
```diff
@@ -379,4 +421,33 @@ class HiveThriftServer2Suite extends FunSuite with Logging {
+        assert(!html.contains("random data that should not be present"))
+
+        // check whether statements exists
+        queries.foreach{ line =>
```
Space before `{`.

liancheng commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suite.scala:
```diff
@@ -70,6 +73,34 @@ class HiveThriftServer2Suite extends FunSuite with Logging {
     port
   }
```
`randomListeningPort` needs to be replaced by something like this:

```scala
  def randomListeningPorts(n: Int) = {
    val sockets = Array.fill(n)(new ServerSocket(0))
    val ports = sockets.map(_.getLocalPort)
    sockets.foreach(_.close())
    ports
  }
```

Because now we need to assign two ports (server and UI).

liancheng commented:
ok to test

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala:
```diff
@@ -28,6 +28,7 @@ import org.apache.spark.annotation.DeveloperApi
 import org.apache.spark.sql.hive.HiveContext
 import org.apache.spark.sql.hive.thriftserver.ReflectionUtils._
 import org.apache.spark.scheduler.{SparkListenerApplicationEnd, SparkListener}
+import org.apache.spark.sql.hive.thriftserver.ui.ThriftServerTab
```
Seems that imports are not ordered properly in this file...

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLSessionManager.scala:
```diff
@@ -50,6 +51,7 @@ private[hive] class SparkSQLSessionManager(hiveContext: HiveContext)
   }
 
   override def closeSession(sessionHandle: SessionHandle) {
+    SparkSQLEnv.sqlEventListener.onDisconnected(super.getSession(sessionHandle))
```
I know that overriding `openSession` needs to move `SparkSQLSessionManager` into Hive shim, so it's not done here. But we should document it and explain when and where `onConnected` is called.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerTab.scala:
```diff
@@ -0,0 +1,45 @@
+ * Spark Web UI tab that shows statistics of a streaming job.
+ * This assumes the given SparkContext has enabled its SparkUI.
+ */
+private[hive] class ThriftServerTab()
```
`private[thriftserver]`?

SparkQA commented:
[Test build #25219 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25219/consoleFull) for   PR 3946 at commit [`f796a81`](https://github.com/apache/spark/commit/f796a81b384cda66c2c83efe7591b078cf10c9da).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25219 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25219/consoleFull) for   PR 3946 at commit [`f796a81`](https://github.com/apache/spark/commit/f796a81b384cda66c2c83efe7591b078cf10c9da).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ThriftServerEventListener `
  * `class SessionInfo(val session: HiveSession, val startTimestamp: Long) `
  * `class ExecutionInfo(val statement: String, val session: HiveSession, val startTimestamp: Long) `
  * `class ThriftServerUIEventListener(val conf: SparkConf)`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25219/
Test FAILed.

liancheng commented on sql/hive-thriftserver/v0.12.0/src/main/scala/org/apache/spark/sql/hive/thriftserver/Shim12.scala:
```diff
@@ -184,8 +184,10 @@ private[hive] class SparkExecuteStatementOperation(
   }
 
   def run(): Unit = {
+    val sid = UUID.randomUUID().toString
```
I guess we can use `parentSession.getSessionHandle.getSessionId` instead?

tianyi replied:
No, the listener needs a statement id, not a session id, for updating each statement's state.

liancheng commented:
@tianyi I understand that because Hive 0.13.1 made incompatible changes to `SessionManager.openSession`, and you'd like to avoid adding new stuff to the shim layer. However, once a shim is introduced, it's doomed to grow :-) I think we can have a `SparkSQLSessionManagerBase` in the shim layer, which only handles `openSession`, and then make `SparkSQLSessionManager` extend it.

@scwf Would you mind to take a look at this? Particularly the job group related part.

scwf commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,179 @@
+
+import scala.xml.Node
+
+/** Page for Spark Web UI that shows statistics of a streaming job */
```
streaming job?

scwf commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,179 @@
+import org.apache.spark.Logging
+import org.apache.spark.ui.UIUtils._
+import org.apache.spark.ui._
+import org.apache.spark.util.Distribution
```
no use import

scwf commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,179 @@
+
+/** Page for Spark Web UI that shows statistics of a streaming job */
+private[ui] class ThriftServerPage(parent: ThriftServerTab)
+  extends WebUIPage("") with Logging {
```
do not need wrap here since it not exceeds 100

scwf commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,179 @@
+          <td>{formatDurationOption(Some(info.totalTime))}</td>
+          <td>{info.statement}</td>
+          <td>{info.state}</td>
+          <td>{errorMessageCell(detail)}</td>
```
how about also show the groupid here?

tianyi replied:
I think showing the `groupId` will make most of spark user feel confusing about it. @liancheng what's your opinion?

liancheng replied:
I'd vote for showing `groupId` here, should be helpful for debugging.

scwf commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,182 @@
+  var executePlan = ""
+  var detail = ""
+  var state: ExecutionState.Value = ExecutionState.STARTED
+  var groupId = ""
```
how/can user set ```groupId``` ?

scwf replied:
@liancheng, if In future ctrl + c can cancel jobs of this group, so if users do not set groupid(use the default value, a empty string), will ctrl +c cancel jobs of all sessions?

liancheng replied:
All session? Do you mean all executing statement within the current session?

tianyi replied:
@scwf @liancheng , I noticed that the `groupId` in `SparkExecuteStatementOperation` had been removed in  #3718. What's this `groupId` used for?

OopsOutOfMemory commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,179 @@
+    </ul>
+  }
+
+  /** Generate stats of batch jobs of the streaming program */
```
Incorrect comments 
`streaming program`   ?

tianyi replied:
Sorry for that, I will update it asap. Thanks

OopsOutOfMemory commented:
@tianyi 
I suggest to move the `plan info`(see below) into `Statement Cell` as a `+detail top-tip` or `options` whatever you like. The `Detail cell` use to display progress information.
```
==Parsed Logical Plan
 xxxx
==Analyzed Logical Plan
 xxxx
......
```

marmbrus commented:
/cc @JoshRosen

tianyi committed:
fix bugs, remove IP trace due to hive-0.12 not support

tianyi commented:
@JoshRosen , Should I remove all the unit tests for thrift server UI? I found u close all UI tests in #3998

JoshRosen commented:
> Should I remove all the unit tests for thrift server UI? I found u close all UI tests in #3998

AFAIK, we didn't need to enable the UI in those unit tests in older branches since it just led to port contention issues and we had no Thrift-specific tests that relied on the Spark web UI.  In your case, I think we should keep the new tests (more tests are pretty much always better!), but we should see if there's a way to avoid port-contention issues (maybe by using a randomized UI port; there might be other tests that do this).  Barring finding a better approach, you could just revert my change in `master` and leave it in place in the other branches (assuming we're not going to port this change to `branch-1.2` (1.2.1)).

tianyi committed:
fix some bug for hive-0.12

SparkQA commented:
[Test build #25521 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25521/consoleFull) for   PR 3946 at commit [`fe7fa28`](https://github.com/apache/spark/commit/fe7fa2882e601191f337752986827a775eb7b565).
 * This patch merges cleanly.

tianyi commented:
@JoshRosen, I made a random ui port in thrift server test suite to avoid the port-contention issues.

SparkQA commented:
[Test build #25521 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25521/consoleFull) for   PR 3946 at commit [`fe7fa28`](https://github.com/apache/spark/commit/fe7fa2882e601191f337752986827a775eb7b565).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25521/
Test PASSed.

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,178 @@
+  /** Render the page */
+  def render(request: HttpServletRequest): Seq[Node] = {
+    val content =
+      generateBasicStats() ++ <br></br> ++
```
I think that this block would be slightly easier to read if you split it across more lines and had more consistent indentation of the HTML tags, e.g.

```scala
val content = 
   generateBasicStats ++
   <br/>  ++
   <h4>
     Total { ...}
     Total {...}
   </h4> ++
   generateSessionStatsTable ++ 
   [...]

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,178 @@
+
+  /** Generate stats of batch statements of the thrift server program */
+  private def generateSQLStatsTable(): Seq[Node] = {
+    val numBatches = listener.executeList.size
```
Is `batches` the right word here?  Seems like this might be a holdover from the streaming UI?

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,178 @@
+      // scalastyle:off
+      <span onclick="this.parentNode.querySelector('.stacktrace-details').classList.toggle('collapsed')"
+            class="expand-details">
+        +details
```
Space after `+`

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,178 @@
+            class="expand-details">
+        +details
+      </span> ++
+        <div class="stacktrace-details collapsed">
```
I'd have two fewer spaces of indentation here, since the `span` and `div` tags are at the same depth in the DOM.

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:
```diff
@@ -0,0 +1,178 @@
+    val numBatches = listener.sessionList.size
+    val table = if (numBatches > 0) {
+      val dataRows =
+        listener.sessionList.values.toSeq.sortBy(_.startTimestamp).reverse.map(session =>{
```
The `=>{` looks a little strange here; I think it would be cleaner to do `.map { session => ` instead.

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerTab.scala:
```diff
@@ -0,0 +1,45 @@
+ * This assumes the given SparkContext has enabled its SparkUI.
+ */
+private[thriftserver] class ThriftServerTab()
+  extends SparkUITab(getSparkUI(), "ThriftServer") with Logging {
```
The initialization / call order here is a bit confusing to me; why not just accept a SparkContext in the `ThriftServerTab` constructor then pass it onto `SparkUITab`?  Why do we need this `getSparkUI()` method?

JoshRosen replied:
This is fine, I suppose; I guess the only change that we could make would be to take the SparkContext through the constructor rather than using SparkSqlEnv; not sure if that's a big win.

tianyi replied:
You mean like this?
```
private[thriftserver] class ThriftServerTab(sparkContext: SparkContext)
  extends SparkUITab(getSparkUI(sparkContext), "ThriftServer") with Logging {

  val parent = getSparkUI(sparkContext)
  val listener = SparkSQLEnv.sqlEventListener

  attachPage(new ThriftServerPage(this))
  parent.attachTab(this)
}

private[thriftserver] object ThriftServerTab {
  def getSparkUI(sparkContext: SparkContext): SparkUI = {
    sparkContext.ui.getOrElse {
      throw new SparkException("Parent SparkUI to attach this tab to not found!")
    }
  }
}
```

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,204 @@
+  /**
+   * Called when a statement started to run.
+   */
+  def onStatementStart(id: String, session: HiveSession, statement: String) { }
```
There should be a blank line after this line (same for the other methods).

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,204 @@
+    val startTimestamp: Long,
+    val ip: String) {
+  val sessionID = session.getSessionHandle.getSessionId.toString
+  val userName = if(session.getUserName == null) "UNKNOWN" else session.getUserName
```
Space after `if`.

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,204 @@
+  var sessionList = new HashMap[SessionHandle, SessionInfo]
+  var executeList = new HashMap[String, ExecutionInfo]
+  val retainedStatements =
+    conf.getInt("spark.thriftserver.ui.retainedStatements", DEFAULT_RETAINED_STATEMENTS)
```
If you're going to use this strategy to limit memory usage, check out the memory leak prevention tests that I wrote for JobProgressListener: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala#L85.  The main idea was to partition the state according to how it grows / shrinks, then write some general assertions.  This may be overkill for what you're doing here, but I just thought that I should mention that approach since it's a nice testing strategy.

JoshRosen replied:
Actually, here's a better reference: #3372

JoshRosen commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,204 @@
+  private def trimExecutionIfNecessary() = synchronized {
+    if (executeList.size > retainedStatements) {
+      val toRemove = math.max(retainedStatements / 10, 1)
+      executeList.toList.sortWith(compareExecutionDesc).take(toRemove).foreach { s =>
```
I think you can use `sortBy(_._2.startTimestamp)` instead of having to define a custom comparison function.

JoshRosen commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suite.scala:
```diff
@@ -384,4 +388,32 @@ class HiveThriftServer2Suite extends FunSuite with Logging {
     }
   }
+
+  test("SPARK-5100 monitor page") {
```
Note that there's a UISeleniumSuite test (currently disabled) for writing much more complex web UI tests for Spark.  Probably overkill here, but just wanted to point it out.

tianyi replied:
@JoshRosen, I have talked with @liancheng about the UISeleniumSuite. I did not add more complex web UI tests because that we worried about the test costs too much time.

JoshRosen commented:
Left a few drive-by comments with pointers to some web UI things.  Overall, this looks substantially similar to the streaming UI, so therefore the changes seem fair to me.  Bumping back to @marmbrus and @liancheng for review of the SQL changes.

Actually, one possible concern: can you start / stop multiple SQLContexts with the same SparkContext?  If so, this might lead to duplicate web UI tabs (I think we have this problem in streaming, so we probably have it here, too: https://issues.apache.org/jira/browse/SPARK-2463.  Not sure if we necessarily want to fix that here, though, since we also need to backport that change to older branches.

tianyi commented:
@JoshRosen , thanks for help. 
About your concerned about `duplicate web UI` problem, it wouldn't happen here. The `thrift server ui` is  created in `HiveThriftServer2`, and i think there wouldn't be duplicate `HiveThriftServer2` in the same application.

tianyi commented:
@liancheng @scwf, since the `groupId` was removed in #3718, I can't find connections between a `job` and a `statement execution` anymore. I tried to use `statement` and `SPARK_JOB_DESCRIPTION`, but it would get wrong when the thrift server is executing two same SQL at the same time. Any suggestion?

scwf commented:
>but it would get wrong when the thrift server is executing two same SQL at the same time.

you mean using two beeline to execute sql(must the same sql?) at same time will get error? what error you get?

tianyi commented:
I mean I can't match a job and a execution via statement, if there are two same sql running



> On Jan 16, 2015, at 11:21, Fei Wang <notifications@github.com> wrote:
> 
> but it would get wrong when the thrift server is executing two same SQL at the same time.
> 
> you mean using two beeline to execute sql(must the same sql?) at same time will get error? what error you get?
> 
> —
> Reply to this email directly or view it on GitHub.
>

tianyi committed:
fix some issues

SparkQA commented:
[Test build #25733 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25733/consoleFull) for   PR 3946 at commit [`14a461d`](https://github.com/apache/spark/commit/14a461dc3dc05b66bd8f6c4027e4c1a39a84d90d).
 * This patch **does not merge cleanly**.

tianyi commented:
rebased from latest master.

SparkQA commented:
[Test build #25735 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25735/consoleFull) for   PR 3946 at commit [`daed3d1`](https://github.com/apache/spark/commit/daed3d126a5112d9e4e94fac7592ff804775ec05).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25735 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25735/consoleFull) for   PR 3946 at commit [`daed3d1`](https://github.com/apache/spark/commit/daed3d126a5112d9e4e94fac7592ff804775ec05).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25735/
Test FAILed.

tianyi committed:
fix shim12 rebase issue; make session id as the default groupId

SparkQA commented:
[Test build #25736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25736/consoleFull) for   PR 3946 at commit [`fb507df`](https://github.com/apache/spark/commit/fb507df555db0084ea7d91ae8a7167d0164480c0).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25733 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25733/consoleFull) for   PR 3946 at commit [`14a461d`](https://github.com/apache/spark/commit/14a461dc3dc05b66bd8f6c4027e4c1a39a84d90d).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25733/
Test PASSed.

SparkQA commented:
[Test build #25736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25736/consoleFull) for   PR 3946 at commit [`fb507df`](https://github.com/apache/spark/commit/fb507df555db0084ea7d91ae8a7167d0164480c0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25736/
Test PASSed.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLSessionManager.scala:
```diff
@@ -30,26 +29,29 @@ import org.apache.spark.sql.hive.thriftserver.server.SparkSQLOperationManager
 
   override def init(hiveConf: HiveConf) {
-    setSuperField(this, "hiveConf", hiveConf)
+
```
Remove this empty line.

liancheng commented on sql/hive-thriftserver/v0.13.1/src/main/scala/org/apache/spark/sql/hive/thriftserver/Shim13.scala:
```diff
@@ -154,8 +176,18 @@ private[hive] class SparkExecuteStatementOperation(
+      case _ => hiveContext.sparkContext
+        .setJobGroup(parentSession.getSessionHandle.getSessionId.toString, statement)
+        sid
+    }
```
Could be simplified to this:

```scala
val groupId = Option(hiveContext.sparkContext.getLocalProperty(SparkContext.SPARK_JOB_GROUP_ID)).getOrElse(sid)
hiveContext.sparkContext.setJobGroup(groupId, statement)
```

We shouldn't use session ID as group ID, otherwise all statements issued by the current session will be cancelled once the job group is cancelled.

liancheng commented on sql/hive-thriftserver/v0.13.1/src/main/scala/org/apache/spark/sql/hive/thriftserver/Shim13.scala:
```diff
@@ -154,8 +176,18 @@ private[hive] class SparkExecuteStatementOperation(
   }
 
   def run(): Unit = {
+    val sid = UUID.randomUUID().toString 
```
Rename it to `statementId`.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala:
```diff
@@ -0,0 +1,197 @@
+  /**
+   * Called when a session created.
+   */
+  def onSessionCreated(ip: String, session: HiveSession) { }
```
Nit: ` { }` => `: Unit = {}`

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIService.scala:
```diff
@@ -73,15 +73,15 @@ private[hive] class SparkSQLCLIService(hiveContext: HiveContext)
 
 private[thriftserver] trait ReflectedCompositeService { this: AbstractService =>
-  def initCompositeService(hiveConf: HiveConf) {
+  def initCompositeService(hiveConf: HiveConf, level: Int = 2) {
```
The `level` argument looks really evil, but I've no idea how to get rid of it right now...

liancheng replied:
I'd suggest moving the whole `SparkSQLSessionManager` class into the shim layer for now. I'm planning to add a reflection based utility that can call 0.12.0 / 0.13.1 APIs adaptively and remove most contents out of the shim layer later.

tianyi committed:
move SparkSQLSessionManager to Shim.
use statementId as groupId to support multiple jobs for one sql.
support multiple jobs for one sql in ThriftServerPage.

SparkQA commented:
[Test build #25792 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25792/consoleFull) for   PR 3946 at commit [`32332cb`](https://github.com/apache/spark/commit/32332cb7830fbf108c6697fb66929a28ee914b3a).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25792 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25792/consoleFull) for   PR 3946 at commit [`32332cb`](https://github.com/apache/spark/commit/32332cb7830fbf108c6697fb66929a28ee914b3a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25792/
Test PASSed.

scwf commented:
`group id` is look ok to me. now we create a uuid for each statement, and save it to EventListener, later  we can make a follow up PR to add the logical that when user presses ctrl-c, use the group id of the statement to cancel the job. Maybe we need add a `onStatementStop` interface to EventListener for that.

liancheng commented:
Actually the CTRL-C part resides in Beeline. Beeline does implemented a signal handler for this, but the registration code is commented out, don't know why...

tianyi committed:
combine ThriftServerUIEventListener with exists listener in HiveThriftServer2

SparkQA commented:
[Test build #26085 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26085/consoleFull) for   PR 3946 at commit [`fd35262`](https://github.com/apache/spark/commit/fd352621c215cf39efb6a3d289d8659366b02c53).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26085 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26085/consoleFull) for   PR 3946 at commit [`fd35262`](https://github.com/apache/spark/commit/fd352621c215cf39efb6a3d289d8659366b02c53).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26085/
Test PASSed.

marmbrus commented:
ping @liancheng

tianyi commented:
Since there is a big difference between this PR and the latest master branch, so I have created a new PR #5730 for this feature. 
So this PR should be closed.
